bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java

```diff
@@ -0,0 +1,253 @@
+                    collection = new ArrayList<>();
+                } else if (Set.class.isAssignableFrom(type)) {
+                    collection = new HashSet<>();
+                } else {
+                    LOGGER.warn("Skipping field '{}', only List and Set is supported as target Collection", fieldName);
+                    continue;
+                }
+                for (final Object it : (Collection<?>) value) {
+                    final Object normalized = valueAs(it, innerClass);
+                    if (normalized == null) {
+                        continue;
+                    }
+                    collection.add(normalized);
+                }
+                value = collection;
+            }
+
+            try {
+                value = valueAs(value, type);
+                if (value == null) {
+                    continue;
+                }
+                LOGGER.trace("Setting value ({}) {} to field '{}' in configuration class {}", type.getSimpleName(),
+                        value, fieldName, configurationClass.getName());
+                field.setAccessible(true);
+                field.set(configuration, value);
+            } catch (SecurityException | IllegalArgumentException | IllegalAccessException ex) {
+                LOGGER.warn("Could not set field value for field '{}': {}", fieldName, ex.getMessage(), ex);
+            }
+        }
+        return configuration;
+    }
+
+    /**
+     * Return fields of the given class as well as all super classes.
+     *
+     * @param clazz The class
+     * @return A list of Field objects
+     */
+    private static List<Field> getAllFields(Class<?> clazz) {
+        List<Field> fields = new ArrayList<>();
+        for (Class<?> superclazz = clazz; superclazz != null; superclazz = superclazz.getSuperclass()) {
+            fields.addAll(Arrays.asList(superclazz.getDeclaredFields()));
+        }
+        return fields;
+    }
+
+    /**
+     * Convert a value to a given type or return default value
+     *
+     * @param value input value or String representation of that value
+     * @param type desired target class
+     * @param defaultValue default value to be used if conversion fails or input value is null
+     * @return the converted value
+     */
+    public static <T> T valueAsOrElse(@Nullable Object value, Class<T> type, T defaultValue) {
+        return Objects.requireNonNullElse(valueAs(value, type), defaultValue);
+    }
+
+    /**
+     * Convert a value to a given type
+     *
+     * @param value input value or String representation of that value
+     * @param type desired target class
+     * @return the converted value or null if conversion fails or input value is null
+     */
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    public static <T> @Nullable T valueAs(@Nullable Object value, Class<T> type) {
+        // make sure primitives are converted to their respective wrapper class
```

Comment:
Should we add an early return if the` value` already is an instance of the requested `type`?